Sources/web3swift/Transaction/TransactionOptions.swift

```diff
@@ -0,0 +1,288 @@
+    ///
+    /// Can be nil if one reads the information from the blockchain.
+    public var from: EthereumAddress?
+
+    public enum GasLimitPolicy {
+        case automatic
+        case manual(BigUInt)
+        case limited(BigUInt)
+        case withMargin(Double)
+    }
+    public var gasLimit: GasLimitPolicy?
+
+    public enum GasPricePolicy {
+        case automatic
+        case manual(BigUInt)
+        case withMargin(Double)
+    }
+
+    public var gasPrice: GasPricePolicy?
+
+    // new gas parameters for EIP-1559 support
+    public enum MaxFeePerGasPolicy {
+        case automatic
+        case manual(BigUInt)
+    }
+    public var maxFeePerGas: MaxFeePerGasPolicy?
+
+    public enum MaxPriorityFeePerGasPolicy {
+        case automatic
+        case manual(BigUInt)
+    }
+    public var maxPriorityFeePerGas: MaxPriorityFeePerGasPolicy?
+
+    /// The value transferred for the transaction in wei, also the endowment if it’s a contract-creation transaction.
+    public var value: BigUInt?
+
+    public enum NoncePolicy {
+        case pending
+        case latest
+        case manual(BigUInt)
+    }
+
+    public var nonce: NoncePolicy?
+
+    public enum CallingBlockPolicy {
+        case pending
+        case latest
+        case exactBlockNumber(BigUInt)
+
+        var stringValue: String {
+            switch self {
+            case .pending:
+                return "pending"
+            case .latest:
+                return "latest"
+            case .exactBlockNumber(let number):
+                return String(number, radix: 16).addHexPrefix()
+            }
+        }
+    }
+
+    public var callOnBlock: CallingBlockPolicy?
+
+    public var accessList: [AccessListEntry]?
+
+    // default options
+    // used by all routines here that fall back to defaults
+    public static var defaultOptions: TransactionOptions {
+        var opts = TransactionOptions()
+        opts.type = .legacy
+        opts.gasLimit = .automatic
+        opts.gasPrice = .automatic
+        opts.maxFeePerGas = .automatic
+        opts.maxPriorityFeePerGas = .automatic
+        opts.nonce = .pending
+        opts.callOnBlock = .pending
+        return opts
+    }
+
+    public func resolveNonce(_ suggestedByNode: BigUInt) -> BigUInt {
+        guard let noncePolicy = self.nonce else { return suggestedByNode }
+        switch noncePolicy {
+        case .pending, .latest:
+            return suggestedByNode
+        case .manual(let value):
+            return value
+        }
+    }
+
+    public func resolveGasPrice(_ suggestedByNode: BigUInt) -> BigUInt {
+        guard let gasPricePolicy = self.gasPrice else { return suggestedByNode }
+        switch gasPricePolicy {
+        case .automatic:
+            return suggestedByNode
+        case .manual(let value):
+            return value
+        case .withMargin:
+            return suggestedByNode
+        }
+    }
+
+    public func resolveGasLimit(_ suggestedByNode: BigUInt) -> BigUInt {
```

Comment:
good suggestion, I don't think it would be required after combining the types to use a common gasPolicy type. gasLimit I think needs to stay separate, as it is handled a little differently, but certainly gasPrice, maxFeePerGas, and maxPriorityFeePerGas can share a common bit of code.

Reply:
Actually on 2nd look that appears to be more difficult than it seemed at first, beyond my current level anyway. Any suggestions on how to accomplish this?

Reply:
Maybe I'm misunderstanding again (probably). I don't see how a generic solves anything here. The input and output are always of type `BigUInt`. The only other component is the selector for which property to resolve, here selection is done via method name. Unless we create a coding key enum, and then basically put a big `switch/case` to select and process each property in the method, I see no real advantage over the current implementation.

Reply:
Yes all the "gasXXX" resolvers return a `BigUInt`. The internal fields just wrap a BigUInt into a "gasPolicy" enum type, or `.automatic` in which case the resolver returns `suggestedByNode` which is the value passed into the resolver function.

Reply:
As far as I know you cannot access a struct by "key" like that.... this is not the same scenario as`decodeHex` where you have a keyed dictionary you are pulling from.

Reply:
I should probably add, that combining these, even if possible, really can't happen in this pass, as this is inherited public API from the existing code base. The only real additional I have here are the addition of the resolvers for the new gas fields.

After further thought, I'm no leaning toward moving `TransactionOptions` back to `Web3+Options` and letting it live there for interacting with `Web3+Transaction`. And then replacing it with a new `TransactionParameters` struct that works at the `EthereumTransaction` level. The more I think about ti, the more problems it seems to solve. Doing this will also reduce the footprint of code I've had to modify outside of `EthereumTransaction` which is a good thing.

Reply:
Actually you' right, i've missed that.

But just incase i've found some nice hack for that, check it https://newbedev.com/how-to-loop-over-struct-properties-in-swift

But this is not to apply here right now.

Reply:
Ah yes mirroring, yeah I was aware of that technique, but felt it starts to obfuscate things a bit too much with all the code you end up wrapping everything in. Not sure the ends justify the means.  Either way, not a problem for right now.

This file is going to go away and return to it's original name/location of `Web3/Web3+options` with only the additions of the new fields added. (and existing API matched resolvers) I think we can look at how to update the resolvers in the gas related PR's, as that is where they ultimately get used.

Reply:
There's not about mirroring, there's a lack of type catsing there, look at the end of the link: there's a hack to encode struct to json and read it like a dict)

Reply:
Ah, yeah but it's terribly inefficient as you are basically converting your structure into a JSON encoding. In that case we could have a slightly more efficient helper-function that creates a dictionary of all the "resolvable" parameters for us. The second challenge then comes from converting the value part of the key/value pair back to the correct `policy` type so it can be acted upon. If they all had the same base type, this would be much easier, but I don't see how we can do that in this case, and ensure that each parameter only uses the cases applicable to it.

As it is we have the following possible maps:
`nonce:` has `.pending`, `.latest`, and `.manual(BigUInt)`
`gasLimit` has `.automatic`, `.manual(BigUInt)`, `withMargin(Double)`, and `.limited(BigUInt)`
`gasPrice` has `.automatic`, `.manual(BigUInt)`, and `withMargin(Double)`
`maxFeePerGas` and `maxPriorityFeePerGas` have `.automatic` and `.manual(BigUInt)` **

**note: These were only put in there as placeholders really, I had not gotten to the part of thinking through what would be appropriate for these values. So it may make some sense for them to share the same options as `gasPrice`. Really depends on how it is going to be handled down the line. Looks like most of the current handling of this is done in `Web3+MutatingTransaction`